test/modules/TestResultCoordinator/Reports/EdgeHubRestartTest/EdgeHubRestartDirectMethodReportGenerator.cs

```diff
@@ -0,0 +1,304 @@
+            this.SenderTestResults = Preconditions.CheckNotNull(senderTestResults, nameof(senderTestResults));
+            this.ReceiverTestResults = Preconditions.CheckNotNull(receiverTestResults, nameof(receiverTestResults));
+            this.PassableEdgeHubRestartPeriod = passableEdgeHubRestartPeriod;
+        }
+
+        internal string TrackingId { get; }
+
+        internal EdgeHubRestartDirectMethodReportMetadata Metadata { get; }
+
+        internal ITestResultCollection<TestOperationResult> SenderTestResults { get; }
+
+        internal ITestResultCollection<TestOperationResult> ReceiverTestResults { get; }
+
+        internal TimeSpan PassableEdgeHubRestartPeriod { get; }
+
+        public async Task<ITestResultReport> CreateReportAsync()
+        {
+            Logger.LogInformation($"Generating report: {nameof(EdgeHubRestartDirectMethodReport)} for [{this.Metadata.SenderSource}] and [{this.Metadata.ReceiverSource}]");
+
+            bool isPassing = true;
+            long previousSeqNum = 0;
+            long passedDirectMethodCount = 0;
+
+            // Value: (source, numOfResults)
+            Dictionary<string, ulong> resultCount = new Dictionary<string, ulong>()
+            {
+                { nameof(this.SenderTestResults), 0ul },
+                { nameof(this.ReceiverTestResults), 0ul }
+            };
+
+            // Value: (completedStatusCode, DirectMethodCompletedTime - EdgeHubRestartedTime)
+            Dictionary<HttpStatusCode, List<TimeSpan>> completedStatusHistogram = new Dictionary<HttpStatusCode, List<TimeSpan>>();
+
+            bool hasSenderResult = await this.SenderTestResults.MoveNextAsync();
+            bool hasReceiverResult = await this.ReceiverTestResults.MoveNextAsync();
+
+            while (hasSenderResult && hasReceiverResult)
+            {
+                this.ValidateResult(
+                    this.SenderTestResults.Current,
+                    this.Metadata.SenderSource,
+                    TestOperationResultType.EdgeHubRestartDirectMethod.ToString());
+
+                this.ValidateResult(
+                    this.ReceiverTestResults.Current,
+                    this.Metadata.ReceiverSource,
+                    TestOperationResultType.DirectMethod.ToString());
+
+                // Both sender & receiver have their dm results
+                resultCount[nameof(this.SenderTestResults)]++;
+                resultCount[nameof(this.ReceiverTestResults)]++;
+
+                // Adjust seqeunce number from both source to be equal before doing any comparison
+                EdgeHubRestartDirectMethodResult senderResult = JsonConvert.DeserializeObject<EdgeHubRestartDirectMethodResult>(this.SenderTestResults.Current.Result);
+                DirectMethodTestResult receiverResult = JsonConvert.DeserializeObject<DirectMethodTestResult>(this.ReceiverTestResults.Current.Result);
+
+                long senderSeqNum = this.ConvertStringToLong(senderResult.SequenceNumber);
```

Comment:
Please have a PR to fix SequenceNubmer in DirectMethodTestResult to use uint instead of string.